ibis/pandas/udf.py

```diff
@@ -211,130 +136,6 @@ def nullable(datatype):
-        the signature of a UDF/UDAF.
-    pin : Optional[int]
-        If this is not None, pin the `pin`-th argument type to `klass`
-    klass : Union[Type[pd.Series], Type[SeriesGroupBy]]
-        The pandas object that every argument type should contain
-
-    Returns
-    -------
-    Tuple[Type]
-        A tuple of types appropriate for use in a multiple dispatch signature.
-
-    Examples
-    --------
-    >>> from pprint import pprint
-    >>> import pandas as pd
-    >>> from pandas.core.groupby import SeriesGroupBy
-    >>> import ibis.expr.datatypes as dt
-    >>> input_type = [dt.string, dt.double]
-    >>> sig = udf_signature(input_type, pin=None, klass=pd.Series)
-    >>> pprint(sig)  # doctest: +ELLIPSIS
-    ((<class '...Series'>, <... '...str...'>, <... 'NoneType'>),
-     (<class '...Series'>,
-      <... 'float'>,
-      <... 'numpy.floating'>,
-      <... 'NoneType'>))
-    >>> not_nullable_types = [
-    ...     dt.String(nullable=False), dt.Double(nullable=False)]
-    >>> sig = udf_signature(not_nullable_types, pin=None, klass=pd.Series)
-    >>> pprint(sig)  # doctest: +ELLIPSIS
-    ((<class '...Series'>, <... '...str...'>),
-     (<class '...Series'>,
-      <... 'float'>,
-      <... 'numpy.floating'>))
-    >>> sig0 = udf_signature(input_type, pin=0, klass=SeriesGroupBy)
-    >>> sig1 = udf_signature(input_type, pin=1, klass=SeriesGroupBy)
-    >>> pprint(sig0)  # doctest: +ELLIPSIS
-    (<class '...SeriesGroupBy'>,
-     (<class '...SeriesGroupBy'>,
-      <... 'float'>,
-      <... 'numpy.floating'>,
-      <... 'NoneType'>))
-    >>> pprint(sig1)  # doctest: +ELLIPSIS
-    ((<class '...SeriesGroupBy'>,
-      <... '...str...'>,
-      <... 'NoneType'>),
-     <class '...SeriesGroupBy'>)
-    """
-    nargs = len(input_type)
-
-    if not nargs:
-        return ()
-
-    if nargs == 1:
-        (r,) = input_type
-        result = (klass,) + rule_to_python_type(r) + nullable(r)
-        return (result,)
-
-    return tuple(
-        klass
-        if pin is not None and pin == i
-        else ((klass,) + rule_to_python_type(r) + nullable(r))
-        for i, r in enumerate(input_type)
-    )
-
-
-def parameter_count(funcsig):
-    """Get the number of positional-or-keyword or position-only parameters in a
-    function signature.
-
-    Parameters
-    ----------
-    funcsig : inspect.Signature
-        A UDF signature
-
-    Returns
-    -------
-    int
-        The number of parameters
-    """
-    return sum(
-        param.kind in {param.POSITIONAL_OR_KEYWORD, param.POSITIONAL_ONLY}
-        for param in funcsig.parameters.values()
-        if param.default is Parameter.empty
-    )
-
-
-def valid_function_signature(input_type, func):
-    """Check that the declared number of inputs (the length of `input_type`)
-    and the number of inputs to `func` are equal.
-
-    Parameters
-    ----------
-    input_type : List[DataType]
-    func : callable
```

Comment:
so most of this was duplicated code?

Reply:
Yes these code are either moved to `ibis.udf.vectorized` or to `pre_execute_reduction_udf` in this file